From: microservice-base scaffold (google_vcpkg variant)
Subject: [PATCH] Work around gcc 13.1.0 ICE in PerCpu NSDMI

gcc 13.1.0 (Qt 6.x's bundled MinGW kit) crashes with an internal
compiler error when instantiating
  std::unique_ptr<T[]> data_{new T[shards_]}
as a non-static data member initializer inside a class template.
Fixed in gcc 13.3+ but Qt's installer ships exactly 13.1.0.

Move the array allocation into the ctor mem-initializer list -
semantically identical, takes a different front-end path that doesn't ICE.

--- a/src/core/util/per_cpu.h
+++ b/src/core/util/per_cpu.h
@@ -89,7 +89,9 @@ class PerCpu {
  public:
   // Options are not defaulted to try and force consideration of what the
   // options specify.
-  explicit PerCpu(PerCpuOptions options) : shards_(options.Shards()) {}
+  explicit PerCpu(PerCpuOptions options)
+      : shards_(options.Shards()),
+        data_(std::unique_ptr<T[]>(new T[shards_])) {}
 
   T& this_cpu() { return data_[sharding_helper_.GetShardingBits() % shards_]; }
 
@@ -101,7 +103,7 @@ class PerCpu {
  private:
   PerCpuShardingHelper sharding_helper_;
   const size_t shards_;
-  std::unique_ptr<T[]> data_{new T[shards_]};
+  std::unique_ptr<T[]> data_;
 };
 
 }  // namespace grpc_core
